Bug 1766907: Add proxy support for CVO.

patrickdillon opened the pull request:
This PR adds proxy support to the CVO for verifying signatures. It centralizes some of the proxy logic used in Cincinnati to be shared. More significantly, it restructures the releaseVerifier struct to be able to use an up-to-date proxy. The releaseVerifier is created before a watch has been established on the cluster proxy. This PR ensures that the releaseVerifier gets the proxy after it is available to the operator and that the proxy settings are updated if there are any changes to the configuration.

Fixes BZ-1766907.

patrickdillon commented:
/retest

patrickdillon commented:
/bugzilla refresh

abhinavdahiya commented on pkg/cvo/cvo.go:
```diff
@@ -652,3 +661,28 @@ func (optr *Operator) defaultPreconditionChecks() precondition.List {
 }
+
+func (optr *Operator) getHTTPClientTransport() (http.RoundTripper, error) {
+	proxyURL, cmNameRef, err := optr.getHTTPSProxyURL()
```
https://github.com/openshift/cluster-version-operator/blob/59601a89bb7a65cf2fbaa4697d5142091f68534f/pkg/cvo/availableupdates.go#L44-L53

does similar, so maybe both can use common method.

wking replied:
This is now centralized in `getTransportOpts`.

abhinavdahiya commented on pkg/cvo/cvo.go:
```diff
@@ -652,3 +661,28 @@ func (optr *Operator) defaultPreconditionChecks() precondition.List {
 	}
 }
+
+func (optr *Operator) getHTTPClientTransport() (http.RoundTripper, error) {
```
should all https clients in cvo pkg use this? because the name most definitely implies this.

abhinavdahiya commented on pkg/verify/verify.go:
```diff
@@ -74,7 +72,7 @@ var Reject Interface = rejectVerifier{}
 // store and the lookup order is internally defined.
 //
-func LoadFromPayload(update *payload.Update) (Interface, error) {
+func LoadFromPayload(update *payload.Update, transport http.RoundTripper) (Interface, error) {
```
i think passing the entire transport is not get an API for this function.

maybe we can think about passing a little more explicit item? transport.Config maybe, or the tlsConfig

jcpowermac replied:
I would figure you would need to pass both the `tlsConfig` and `proxyURL`.
Something like what is in: 
https://github.com/jcpowermac/cluster-version-operator/blob/master/pkg/cvo/availableupdates.go

Then create the http client:
https://github.com/jcpowermac/cluster-version-operator/blob/d4e138fac270b557fb9d2592d000de40fcd9affd/pkg/cincinnati/cincinnati.go#L64-L72

abhinavdahiya commented:
/cc @smarterclayton

patrickdillon commented:
/cc @jcpowermac

patrickdillon commented:
/retest

patrickdillon commented:
/test images

patrickdillon commented:
/test images

sdodson commented:
/retest

patrickdillon commented:
/retest

patrickdillon commented:
/retest

patrickdillon commented:
/test images

patrickdillon commented:
/test images

patrickdillon commented:
@abhinavdahiya I have tested this and verified that `storage.googleapis.com` is now read in the proxy logs which is an indication that this PR is working. 

I have also updated the PR description to reflect that we need to take into account that the `releaseVerifier` is created before the operator has a watch on the cluster proxy, so we cannot read the settings for the proxy when creating the releaseVerifier. We also need to make sure to update the client if the proxy changes. 

The current implementation addresses these issues by creating an interface implemented by the operator itself which will pull the current proxy settings and create a new httpClient whenever the releaseVerifier needs to verify signatures.

abhinavdahiya commented on pkg/verify/verify.go:
```diff
@@ -74,7 +72,7 @@ var Reject Interface = rejectVerifier{}
 // store and the lookup order is internally defined.
 //
-func LoadFromPayload(update *payload.Update) (Interface, error) {
+func LoadFromPayload(update *payload.Update, clientBuilder clientBuilder) (Interface, error) {
```
I don't think an argument of type `package private` is a good pattern for `public` function

patrickdillon replied:
Updated to export ClientBuilder interface type/

abhinavdahiya commented on pkg/cvo/cvo.go:
```diff
@@ -224,7 +226,7 @@ func (optr *Operator) InitializeFromPayload(restConfig *rest.Config, burstRestCo
 
 	// attempt to load a verifier as defined in the payload
-	verifier, err := verify.LoadFromPayload(update)
+	verifier, err := verify.LoadFromPayload(update, optr)
```
instead of passing the whole optr object, can we pass a small type that just does HTTPClient .. 
you can `optr.verifyClientBuilder` or some such.

patrickdillon replied:
I think we can do that if we store pointers to the listers for proxy and trust bundle in the new type.

patrickdillon replied:
Created a wrapper struct to hold the function.

abhinavdahiya commented on pkg/verify/verify.go:
```diff
@@ -150,7 +142,19 @@ var validReleaseDigest = regexp.MustCompile(`^[a-zA-Z0-9:]+$`)
+}
+
+type clientBuilder interface {
+	GetHTTPClient() (*http.Client, error)
```
nit: `GetHTTPClient` -> `HTTPClient`

patrickdillon replied:
done

patrickdillon commented:
This is ready for another round of reviews. Suggestions have been incorporated and tested (I can see requests to storage.googleapis.com in proxy).

abhinavdahiya commented on pkg/verify/verify.go:
```diff
@@ -256,7 +261,11 @@ func (v *releaseVerifier) Verify(ctx context.Context, releaseDigest string) erro
+			client, err := v.clientBuilder.HTTPClient()
+			if err != nil {
+				return err
+			}
```
I think we are createing an http client for each store, maybe we can do it once per Verify call ?

patrickdillon replied:
Absolutely. I moved client creation out of the for loop.

patrickdillon commented:
This is ready for another round of reviews. Let me know if there are certain situations I can test. 

With this patch I can see requests to storage.googleapis.com on the proxy (which aren't there otherwise). Those requests get 404 responses but I believe that is to be expected.

abhinavdahiya commented:
/lgtm

abhinavdahiya commented:
/cherrypick release-4.2
